addons/stock/models/stock_move.py

```diff
@@ -459,11 +459,24 @@ def _compute_json_forecast(self):
                 else:
                     move.json_forecast = json.dumps({'expectedDate': None})
 
-    def _set_date_deadline(self):
+    def _set_date_deadline(self, new_deadline):
         # Handle the propagation of `date_deadline` fields (up and down stream - only update by up/downstream documents)
+        already_propagate_ids = self.env.context.get('date_deadline_propagate_ids', set()) | set(self.ids)
+        self = self.with_context(date_deadline_propagate_ids=already_propagate_ids)
         for move in self:
-            move_linked = (move.move_dest_ids | move.move_orig_ids).filtered(lambda m: m.state not in ('done', 'cancel'))
-            move_linked.filtered(lambda m: m.date_deadline != move.date_deadline).date_deadline = move.date_deadline
+            moves_to_update = (move.move_dest_ids | move.move_orig_ids).filtered(lambda m: m.state not in ('done', 'cancel') and m.id not in already_propagate_ids)
+            if not moves_to_update:
+                continue
+            if not new_deadline:
+                moves_to_update.date_deadline = new_deadline
+                continue
+            if move.date_deadline:
+                delta = move.date_deadline - fields.Datetime.to_datetime(new_deadline)
+            for move_update in moves_to_update:
+                if move_update.date_deadline and delta:
+                    move_update.date_deadline -= delta
+                else:
+                    move_update.date_deadline = new_deadline
```

Comment:
In case of move.date_deadline == False and new_deadline != False, check on move.rule_id.delay to propagate the deadline correctly